src/server/services/home-block.service.ts

```diff
@@ -153,3 +208,101 @@ export const getHomeBlockData = async ({
       return { ...homeBlock, metadata };
   }
 };
+
+export const userHasCustomHomeBlocks = async (userId?: number) => {
+  if (!userId) {
+    return false;
+  }
+
+  const [row]: { exists: boolean }[] = await dbRead.$queryRaw`
+    SELECT EXISTS(
+        SELECT 1 FROM "HomeBlock" hb WHERE hb."userId"=${userId} AND hb."sourceId" IS NULL
+      )
+  `;
+
+  const { exists } = row;
+
+  return exists;
+};
+
+export const upsertHomeBlock = async ({
+  input,
+}: {
+  input: UpsertHomeBlockInput & { userId: number; isModerator?: boolean };
+}) => {
+  const { userId, isModerator, id, metadata, type, sourceId, index } = input;
+
+  if (id) {
+    const homeBlock = await dbRead.homeBlock.findUnique({
+      select: { userId: true },
+      where: { id },
+    });
+
+    if (!homeBlock) {
+      throw throwNotFoundError('Home block not found.');
+    }
+
+    if (userId !== homeBlock.userId && !isModerator) {
+      throw throwAuthorizationError('You are not authorized to edit this home block.');
+    }
+
+    const updated = await dbWrite.homeBlock.updateMany({
+      where: { OR: [{ id }, { sourceId: id }] },
+      data: {
+        metadata,
+        index,
+      },
+    });
+
+    return updated;
+  }
+
+  return dbWrite.homeBlock.create({
+    data: {
+      metadata,
+      type,
+      sourceId,
+      index,
+      userId,
+    },
+  });
+};
+
+export const deleteHomeBlockById = async ({
+  input,
+}: {
+  input: GetByIdInput & { userId: number; isModerator?: boolean };
+}) => {
+  try {
+    const { id, userId, isModerator } = input;
+    const homeBlock = await dbRead.homeBlock.findFirst({
+      // Confirm the homeBlock belongs to the user:
+      where: { id, userId: isModerator ? undefined : userId },
```

Comment:
This also works without the middleware 😛 nice workaround

Reply:
Indeed, I didn't think to protect it API level but service 😬. But might be a good call